Rgbd camera all shape test

kunimatsu-tri opened the pull request:
Added a unit test for rendering all the shapes that `RgbdCamera` can handle at this moment.
This PR also includes fixing a bug that cylinder's orientation in VTK was different from that of Drake.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5382)
<!-- Reviewable:end -->

kunimatsu-tri commented:
+@liangfok feature review please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeF3IwhLsqna9dqspon:b-yklb5l)*
<!-- Sent from Reviewable.io -->

liangfok commented:
@kunimatsu-tri I submitted a refinement to this PR and left a couple comments. PTAL. 

---

Reviewed 11 of 11 files at r1, 1 of 3 files at r2, 1 of 1 files at r3.
Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFGcU0vjgXbKsZhH5x:-KeFGcU0vjgXbKsZhH5y:b5iyiym) (no related file):*
BTW, the `bottle.png` is 593KB. Can it be made smaller?

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 226 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFJ31xTMmDyslnaBOT:-KeFIjANUcFnemQw6HNB:b2gbhhl) ([raw file](https://github.com/robotlocomotion/drake/blob/e1e9317c3e468bef8c5777248d5893fb11859881/drake/systems/sensors/test/rgbd_camera_test.cc#L226)):*
> ```cpp
>         ASSERT_NEAR(color_image.at(u, v)[1], *(color + 1), 1.);  // G
>         ASSERT_NEAR(color_image.at(u, v)[2], *(color + 0), 1.);  // R
>         ASSERT_NEAR(color_image.at(u, v)[3], *(color + 3), 1.);  // A
> ```

BTW, is this using a floating point macro (`ASSERT_NEAR()`) to check two integer types?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeFK6UMfEeQpr4AkcwS:b-8ouoo1)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 2 of 3 files at r2, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFGcU0vjgXbKsZhH5x:-KeFN6JGPZdPDOSGdNN9:b62rjvd) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, the `bottle.png` is 593KB. Can it be made smaller?
</blockquote></details>

I tried with meshlab and reduced from 1.1MB to 593KB.
That was at my best with meshlab and could't find another way.

Are you familiar with doing that?

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 226 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFJ31xTMmDyslnaBOT:-KeFMvyyB10OQmzLSbmW:b4f5kq8) ([raw file](https://github.com/robotlocomotion/drake/blob/e1e9317c3e468bef8c5777248d5893fb11859881/drake/systems/sensors/test/rgbd_camera_test.cc#L226)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is this using a floating point macro (`ASSERT_NEAR()`) to check two integer types?
</blockquote></details>

Oops, that was for uint8_t. will remove `.` in the tolerance section.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFGcU0vjgXbKsZhH5x:-KeFK4jscDWO1uFbxTY5:be99o7m) (no related file):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I tried with meshlab and reduced from 1.1MB to 593KB.
That was at my best with meshlab and could't find another way.

Are you familiar with doing that?
</blockquote></details>

Since this is a `.png`, can we make it smaller using Gimp? https://www.gimp.org/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFGcU0vjgXbKsZhH5x:-KeFPj7lNeYXnJSc_F_x:b-99o90x) (no related file):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since this is a `.png`, can we make it smaller using Gimp? https://www.gimp.org/
</blockquote></details>

This is associated with .obj and I don't know what will happen by doing that, but let me try.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Oops, I made mistake. Let me try again.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeFWlUL1kdvMjVYVFtP:b51b1fi)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments. PTAL.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeFZovCkzIbr8BeDOgB:b-y0j4qv)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending one question.

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 189 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeF_mTRSd07ESKV8OhN:-KeF_mTRSd07ESKV8OhO:bdvndfv) ([raw file](https://github.com/robotlocomotion/drake/blob/43275c46107cde7d6286b695f08a8b50265b8fa7/drake/systems/sensors/test/rgbd_camera_test.cc#L189)):*
> ```cpp
>         ASSERT_NEAR(color_image.at(u, v)[1], 229u, kColorPixelTolerance);
>         ASSERT_NEAR(color_image.at(u, v)[2], 255u, kColorPixelTolerance);
>         ASSERT_NEAR(color_image.at(u, v)[3], 255u, kColorPixelTolerance);
> ```

BTW, could these instead be the following?

```
ASSERT_LT(std::abs(color_image.at(u, v)[3] - 255u), kColorPixelTolerance);
```

I'm still concerned about using a floating-point-specific macro to check non-floating point values.

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#floating-point-macros

Feel free to ignore if my concern is invalid.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeFaV0K3Wuf0aI7G_1q:b-kyod8q)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/test/rgbd_camera_test.cc, line 189 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeF_mTRSd07ESKV8OhN:-KeFeAGj12MHv56rJazO:b-1tnwco) ([raw file](https://github.com/robotlocomotion/drake/blob/43275c46107cde7d6286b695f08a8b50265b8fa7/drake/systems/sensors/test/rgbd_camera_test.cc#L189)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, could these instead be the following?

```
ASSERT_LT(std::abs(color_image.at(u, v)[3] - 255u), kColorPixelTolerance);
```

I'm still concerned about using a floating-point-specific macro to check non-floating point values.

https://github.com/google/googletest/blob/master/googletest/docs/AdvancedGuide.md#floating-point-macros

Feel free to ignore if my concern is invalid.
</blockquote></details>

Oh I've now caught your point and that's true. Thanks for pointing it out!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
+@jwnimmer-tri platform review please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeFgyycM62rKQmicR90:b-k0crfs)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/test/rgbd_camera_test.cc, line 189 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeF_mTRSd07ESKV8OhN:-KeFg7io-zYcLCzwrVr9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/43275c46107cde7d6286b695f08a8b50265b8fa7/drake/systems/sensors/test/rgbd_camera_test.cc#L189)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Oh I've now caught your point and that's true. Thanks for pointing it out!
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 11 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeFxN6Tb9WR-OHR5ZyI:btdtffa) (no related file):*
To confirm my understanding before I read more... this is a "golden wall of numbers" test, yes?  We set up a scene using SDF and OBJ, ask the sensor to render it, and then compare the result to pre-computed files on disk with the rgb and depth images?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeFy2OKQSoFmFXgSGIn:b-6jahs3) (no related file):*
If "golden wall of numbers" test means the sentence below, yes, that is correct.

> We set up a scene using SDF and OBJ, ask the sensor to render it, and then compare the result to pre-computed files on disk with the rgb and depth images?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeFycBSn8bQkt9kzMKw:b-s1hpbh) (no related file):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

If "golden wall of numbers" test means the sentence below, yes, that is correct.

> We set up a scene using SDF and OBJ, ask the sensor to render it, and then compare the result to pre-computed files on disk with the rgb and depth images?
</blockquote></details>

If I change the camera code in such a way that the DUT answers (the Image object produced by the sensor) and the golden answers (color.png, depth.png) disagree, how should I assess whether the new results are correct, and how do I update the answer files on disk if the new result is better?  Does the camera class output new golden files that I can just copy into place?  If a new developer doesn't have a ready answer to those questions, then the test can never evolve, and thus will become a pain point.

If solving those two questions is hard, then maybe a golden-answer test is not what you want here, and that setting up a more focused scenario (e.g., render a plane and check some facts about it) would be good enough?  I see on master that there is some tests started like that already.  What is the purpose of this new test, and can we do it with a parametric assessment, rather than golden answers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeG-jTrbr9CsMlBEgiY:bj3skzi) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If I change the camera code in such a way that the DUT answers (the Image object produced by the sensor) and the golden answers (color.png, depth.png) disagree, how should I assess whether the new results are correct, and how do I update the answer files on disk if the new result is better?  Does the camera class output new golden files that I can just copy into place?  If a new developer doesn't have a ready answer to those questions, then the test can never evolve, and thus will become a pain point.

If solving those two questions is hard, then maybe a golden-answer test is not what you want here, and that setting up a more focused scenario (e.g., render a plane and check some facts about it) would be good enough?  I see on master that there is some tests started like that already.  What is the purpose of this new test, and can we do it with a parametric assessment, rather than golden answers?
</blockquote></details>

Thanks.  I totally understand your point.  It is very annoying to maintain this kind of thing.

Let me explain what this test is meant for.

The test that exists on master is just a flat terrain and there is no other object in the scene except the terrain.  So, I want to have tests for other shapes including primitive shapes(box, sphere and cylinder) and mesh.

Maybe we can separate `AllShapeRenderingTest` into four different tests corresponding to aforementioned each shapes.  And with carefully chosen configurations, namely camera pose, shape pose and size, we can do parametric test for the depth image in the way that developers can easily calculate expected values by hand.  In this case, the mesh object should be more simplistic one like a box rather than the currently used bottle.

For the color image, the pixel values vary depending on the shape surface direction w.r.t. the camera optical axis, so it might be better not to have tests for that rather than golden wall tests.

What would do think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed the comment about "golden wall" test. PTAL.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeJWObUHMIwWqcy8UpL:bzdok95)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 11 files at r1, 6 of 13 files at r7.
Review status: 7 of 9 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeJZpMMt30WfA4PmbJU:b6j8pok) (no related file):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Thanks.  I totally understand your point.  It is very annoying to maintain this kind of thing.

Let me explain what this test is meant for.

The test that exists on master is just a flat terrain and there is no other object in the scene except the terrain.  So, I want to have tests for other shapes including primitive shapes(box, sphere and cylinder) and mesh.

Maybe we can separate `AllShapeRenderingTest` into four different tests corresponding to aforementioned each shapes.  And with carefully chosen configurations, namely camera pose, shape pose and size, we can do parametric test for the depth image in the way that developers can easily calculate expected values by hand.  In this case, the mesh object should be more simplistic one like a box rather than the currently used bottle.

For the color image, the pixel values vary depending on the shape surface direction w.r.t. the camera optical axis, so it might be better not to have tests for that rather than golden wall tests.

What would do think?

</blockquote></details>

Sounds much better, I will take a look.

For RGB, is it truly the case that there is no way to have an easily-understood reference answer?  For example, if the object is pure red (or blue, or etc.) and some kind of global illumination is set to maximum?  Or that some non-RGB colorspace is more easily checked?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeJahvF_ojJAi8jeZcE:b-6berff) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Sounds much better, I will take a look.

For RGB, is it truly the case that there is no way to have an easily-understood reference answer?  For example, if the object is pure red (or blue, or etc.) and some kind of global illumination is set to maximum?  Or that some non-RGB colorspace is more easily checked?
</blockquote></details>

I now think we can do that for flat surfaces, it might be hard for curved surfaces though. Let me try.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeJcfW1NXQFvHVlAbmy:b-brozuj) (no related file):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I now think we can do that for flat surfaces, it might be hard for curved surfaces though. Let me try.
</blockquote></details>

Surely there's some kind of "this pixel is black" vs "this pixel is a bit red" that we could at least make sure the correct pixels in the RGB image were either the object or the background?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeJd9rpiNUQNp2viqb0:b-ywka5u) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Surely there's some kind of "this pixel is black" vs "this pixel is a bit red" that we could at least make sure the correct pixels in the RGB image were either the object or the background?
</blockquote></details>

Oh, that's a good point!  I will try checking whether it belongs to the background or not for curved surfaces.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 75 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJohPEop7V50I-gvS0:-KeJohPFBaHEkeqGXhju:bvoblun) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L75)):*
> ```cpp
>   const size_t last_dot = filepath.find_last_of(".");
>   if (last_dot == std::string::npos) {
>     DRAKE_ASSERT(false);
> ```

This is definitely not good.  If we are going to abort on malformed filenames, then we need to DRAKE_DEMAND at least, so that the check is always armed.

---

*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeJp5WsHsFFgwBOx38n:b-9690ma) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
> ```cpp
>         std::ifstream file_exist(texture_file);
> 
>         if (file_exist) {
> ```

This seems brittle.  Why do we sliently ignore a missing texture?  Is it possible to change the API  or code so that we _know_ if we should expect a texture, and then fail-fast if we didn't find it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 75 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJohPEop7V50I-gvS0:-KeJqCIX0FlnkrVmmhwR:beyqtlu) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L75)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This is definitely not good.  If we are going to abort on malformed filenames, then we need to DRAKE_DEMAND at least, so that the check is always armed.
</blockquote></details>

Thanks for pointing it out. will replace it with DRAKE_DEMAND.

---

*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeJqL-pgMPIQI3ji0hM:b-xnqfgq) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This seems brittle.  Why do we sliently ignore a missing texture?  Is it possible to change the API  or code so that we _know_ if we should expect a texture, and then fail-fast if we didn't find it?
</blockquote></details>

I thought it's worth rendering the mesh with a default color even if the texture file doesn't exist.

I'm assuming a case that users don't have a texture file since their .obj file is converted from e.g. .stl which doesn't support texture.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeJs1IwgqFyyMwwcgkz:b-h4si1u) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

I thought it's worth rendering the mesh with a default color even if the texture file doesn't exist.

I'm assuming a case that users don't have a texture file since their .obj file is converted from e.g. .stl which doesn't support texture.
</blockquote></details>

Sure.  We should still support cases where there's only the obj data.  But _guessing_ whether the user wanted obj-only or obj+png is the defect here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 75 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJohPEop7V50I-gvS0:-KeJsBES4alaATPCstto:b-zcjvte) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L75)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Thanks for pointing it out. will replace it with DRAKE_DEMAND.
</blockquote></details>

Thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeJtdFHdIFwG9o3r2Mn:bsz0qso) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Sure.  We should still support cases where there's only the obj data.  But _guessing_ whether the user wanted obj-only or obj+png is the defect here.
</blockquote></details>

Just to make sure your point, you are saying that the current implementation can not judge whether it's "obj-only" case or "obj+png" case but png file doesn't have an expected name. Is this correct?
 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 11 files at r1, 8 of 13 files at r7.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 72 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJyXGEMqpD2sNOL8XM:-KeJyXGFEQEt41MPCFiE:blekqiv) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L72)):*
> ```cpp
> }
> 
> std::string RemoveFileExtention(const std::string& filepath) {
> ```

BTW, typo: "Extension".

---

*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeJyltFhlgm6wMFMndN:byf9hep) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Just to make sure your point, you are saying that the current implementation can not judge whether it's "obj-only" case or "obj+png" case but png file doesn't have an expected name. Is this correct?
 
</blockquote></details>

Maybe we can add a new type called `DrakeShapes::MESH_WITH_TEXTURE` so users can explicitly indicate that a texture must be found? I worry that including such a change would result in this PR growing too big.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 188 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJz8tdkohLX-omW2Td:-KeJz8tdkohLX-omW2Te:b-fc6eg2) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/rgbd_camera_test.cc#L188)):*
> ```cpp
>       for (int u = 0; u < color_image.width(); u += 20) {
>         // We need kColorPixelTolerance because it is possible to have rendering
>         // errors dependeing on the hardware that the rendering library uses.
> ```

BTW, typo: "depending".

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJzRdYlvMojyPiREEu:-KeJzRdYlvMojyPiREEv:b-6c4m7g) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/rgbd_camera_test.cc#L254)):*
> ```cpp
> // Verifies the rendered box.
> GTEST_TEST(RenderingTest, BoxRenderingTest) {
>   const std::string sdf("/systems/sensors/test/models/box.sdf");
> ```

BTW, meta, should this use `GetDrakePath()` so the unit test can run from any directory?

---

*[drake/systems/sensors/test/models/mesh_box.sdf, line 26 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJzk0x0CF5DFfvcACj:-KeJzk0x0CF5DFfvcACk:brh8w1i) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/models/mesh_box.sdf#L26)):*
> ```
>           <box>
>             <size>1 1 1</size>
> 	  </box>
> ```

BTW, consider changing the tab to spaces.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeKFNglH65lvqsfkPAR:b-3hpqe7) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Maybe we can add a new type called `DrakeShapes::MESH_WITH_TEXTURE` so users can explicitly indicate that a texture must be found? I worry that including such a change would result in this PR growing too big.
</blockquote></details>

`DrakeShapes::MESH_WITH_TEXTURE` approach sounds good to me.

How about leaving a TODO comment that says we will need to distinguish aforementioned two cases when `MESH_WITH_TEXTURE` is implemented for now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeKSFLNNEFtj1BEq9D_:bd9qh67) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

`DrakeShapes::MESH_WITH_TEXTURE` approach sounds good to me.

How about leaving a TODO comment that says we will need to distinguish aforementioned two cases when `MESH_WITH_TEXTURE` is implemented for now?


</blockquote></details>

I would say either a TODO or an issue will be OK, but let's wait for the platform reviewer to weigh in and make the final call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeKWXhJmWf_YhkP2JLj:btfbdju) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

I would say either a TODO or an issue will be OK, but let's wait for the platform reviewer to weigh in and make the final call.
</blockquote></details>

I am OK with resolving this discussion with a TODO.  If that's your choice, please phrase the TODO more like "guessing is bad" than "add a new enum to DrakeShapes".

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJzRdYlvMojyPiREEu:-KeKWnfsHJulh3bYul0t:b-5dk47n) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/rgbd_camera_test.cc#L254)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, meta, should this use `GetDrakePath()` so the unit test can run from any directory?
</blockquote></details>

If you read up above, the 1st argument here _is_ a suffix atop GetDrakePath.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 254 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJzRdYlvMojyPiREEu:-KeKXiZSRi1f1e9WLAMe:b-nxyju2) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/rgbd_camera_test.cc#L254)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If you read up above, the 1st argument here _is_ a suffix atop GetDrakePath.
</blockquote></details>

Excellent. Marking myself satisfied.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[drake/systems/sensors/rgbd_camera.cc, line 315 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJp5WrJhYVw-M3pJgC:-KeKX_hmEDALtMabFm1G:b-4hvi0m) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L315)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am OK with resolving this discussion with a TODO.  If that's your choice, please phrase the TODO more like "guessing is bad" than "add a new enum to DrakeShapes".
</blockquote></details>

Got it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Review status: 5 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/sensors/rgbd_camera.cc, line 72 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJyXGEMqpD2sNOL8XM:-KeKGuBJrPOXdQ7LvM0R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/rgbd_camera.cc#L72)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, typo: "Extension".
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/rgbd_camera_test.cc, line 188 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJz8tdkohLX-omW2Td:-KeKEsG6PTYxyTF6l5SP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/rgbd_camera_test.cc#L188)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, typo: "depending".
</blockquote></details>

Done.

---

*[drake/systems/sensors/test/models/mesh_box.sdf, line 26 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeJzk0x0CF5DFfvcACj:-KeKWxNt-qQL9StGhIS6:b6e6q1o) ([raw file](https://github.com/robotlocomotion/drake/blob/23670ee6b77563d94c9ad58bcb08bfbfe4cff687/drake/systems/sensors/test/models/mesh_box.sdf#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider changing the tab to spaces.
</blockquote></details>

Sorry. Changed my emacs setting to visualize tabs. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-KeFxN6SDD7t7Di4tWvg:-KeLJBYBPoT-QhgMvQFF:bd1z9y0) (no related file):*
<details><summary><i>Previously, kunimatsu-tri (Kunimatsu Hashimoto) wrote…</i></summary><blockquote>

Oh, that's a good point!  I will try checking whether it belongs to the background or not for curved surfaces.
</blockquote></details>

For sphere, calculating pixels belonging to the sphere by hand was a bit hard, so tested image's four corners (obviously outside)  and center (obviously inside) only.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
Addressed all the comments. PTAL. Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeLJ4z7qQCJMQrIxGPw:b-thxh1r)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 5 of 5 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental please
@drake-jenkins-bot linux-clang-experimental-memcheck-lsan please
@drake-jenkins-bot linux-xenial-clang-experimental please
@drake-jenkins-bot linux-xenial-clang-experimental-debug please
@drake-jenkins-bot linux-xenial-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeOduWyZPhMTeg3c0Wl:b-rgn5i0)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
@drake-jenkins-bot linux-clang-experimental-debug please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeOeEvrPdm6f-mXU_v3:bxvg7ss)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 11 files at r1, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
@drake-jenkins-box mac-clang-experimental please
@drake-jenkins-box mac-clang-ninja-experimental please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KePDGlaAVlxbVeZJhc2:b-qxwo2m)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
@drake-jenkins-bot mac-clang-experimental please
@drake-jenkins-bot mac-clang-ninja-experimental please

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KePD_PknPchmiQr8IKo:bg0xcbm)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 11 files at r1, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Oops, sorry, I thought I already signed this.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeZ-EBLS9dtKtYsi-7q:b6lthjo)*
<!-- Sent from Reviewable.io -->

kunimatsu-tri commented:
No problem. Thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5382#-:-KeZ4ahfXBA1xHXhSdIG:btt5nmp)*
<!-- Sent from Reviewable.io -->
